MapleServer2/Types/AdditionalEffects.cs

```diff
@@ -177,6 +250,171 @@ public void RemoveEffect(AdditionalEffect effect, int index, bool sendPacket = t
+    {
+        if (start == -1)
+        {
+            start = Environment.TickCount;
+        }
+
+        for (int skillIndex = 0; skillIndex < trigger.SkillId.Count(); ++skillIndex)
+        {
+            int skillId = trigger.SkillId[skillIndex];
+            int skillLevel = trigger.SkillLevel[skillIndex];
+
+            if (parent.SkillFiredLast.TryGetValue(skillId, out int lastFired))
+            {
+                if (start - lastFired < GetTriggerCooldown(skillId, skillLevel))
+                {
+                    continue;
+                }
+            }
+
+            float probability = trigger.BeginCondition?.Probability ?? 1;
+
+            if (probability != 1 && probability < Random.Shared.NextDouble())
+            {
+                continue;
+            }
+
+            parent.SkillFiredLast[skillId] = start;
+
+            SkillCast skillCast = new(skillId, (short) skillLevel, parentSkill.SkillSn, start, parentSkill)
+            {
+                CasterObjectId = parentSkill.CasterObjectId,
+                Position = Parent.Coord,
+                Interval = trigger.Interval,
+                Duration = parent?.LevelMetadata?.Basic?.DurationTick ?? 100
+            };
+
+            if (trigger.Splash == true)
+            {
+                RegionSkillHandler.HandleEffect(Parent.FieldManager, skillCast, skillCast.SkillAttack.AttackPoint);
+
+                continue;
+            }
+
+            target.AdditionalEffects.AddEffect(new(skillId, skillLevel)
+            {
+                Duration = skillCast.DurationTick(),
+                Source = parentSkill.CasterObjectId,
+                ParentSkill = parentSkill
+            });
+        }
+    }
+
+    public void SkillTrigger(IFieldActor target, SkillCast skillActivated)
+    {
+        foreach (AdditionalEffect effect in Effects)
+        {
+            if (!effect.ListensForSkills)
+            {
+                continue;
+            }
+
+            if (!ShouldEffectBeActive(effect))
+            {
+                continue;
+            }
+
+            foreach (EffectTriggerSkillMetadata skill in effect.LevelMetadata.ConditionSkill)
+            {
+                if (skill.BeginCondition.Owner.EventSkillIDs == null)
+                {
+                    FireTriggerSkill(effect, skill, skillActivated, target, Environment.TickCount);
+
+                    continue;
+                }
+
+                foreach (int skillId in skill.BeginCondition.Owner.EventSkillIDs)
+                {
+                    if (skillId == skillActivated.SkillId)
+                    {
+                        FireTriggerSkill(effect, skill, skillActivated, target, Environment.TickCount);
+                    }
+                }
+            }
+        }
+    }
+
+    public void EffectTrigger(IFieldActor target, AdditionalEffect effectActivated)
```

Comment:
Never used?

Reply:
I haven't implemented any features that use this directly yet, but it exists in the xmls so I set up the infrastructure for it while setting up the one for skill triggers.